sklearnex/cluster/k_means.py

```diff
@@ -1,17 +1,370 @@
+
+    @control_n_jobs(decorated_methods=["fit", "predict", "transform", "fit_transform"])
+    class KMeans(sklearn_KMeans, BaseKMeans):
+        __doc__ = sklearn_KMeans.__doc__
+        n_iter_, inertia_ = None, None
+        labels_, cluster_centers_ = None, None
+
+        if sklearn_check_version("1.2"):
+            _parameter_constraints: dict = {**sklearn_KMeans._parameter_constraints}
+
+        def __init__(
+            self,
+            n_clusters=8,
+            *,
+            init="k-means++",
+            n_init=(
+                "auto"
+                if sklearn_check_version("1.4")
+                else "warn" if sklearn_check_version("1.2") else 10
+            ),
+            max_iter=300,
+            tol=1e-4,
+            verbose=0,
+            random_state=None,
+            copy_x=True,
+            algorithm="lloyd" if sklearn_check_version("1.1") else "auto",
+        ):
+            super().__init__(
+                n_clusters=n_clusters,
+                init=init,
+                max_iter=max_iter,
+                tol=tol,
+                n_init=n_init,
+                verbose=verbose,
+                random_state=random_state,
+                copy_x=copy_x,
+                algorithm=algorithm,
+            )
+
+        def _initialize_onedal_estimator(self):
+            onedal_params = {
+                "n_clusters": self.n_clusters,
+                "init": self.init,
+                "max_iter": self.max_iter,
+                "tol": self.tol,
+                "n_init": self.n_init,
+                "verbose": self.verbose,
+                "random_state": self.random_state,
+            }
+
+            self._onedal_estimator = onedal_KMeans(**onedal_params)
+
+        def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
+            assert method_name == "fit"
+
+            class_name = self.__class__.__name__
+            patching_status = PatchingConditionsChain(f"sklearn.cluster.{class_name}.fit")
+
+            sample_count = _num_samples(X)
+            self._algorithm = self.algorithm
+            supported_algs = ["auto", "full", "lloyd", "elkan"]
+            correct_count = self.n_clusters < sample_count
+            is_data_supported = (
+                _is_csr(X) and daal_check_version((2024, "P", 600))
+            ) or not issparse(X)
+            sample_weight = _check_sample_weight(
+                sample_weight, X, dtype=X.dtype if hasattr(X, "dtype") else None
+            )
+
+            patching_status.and_conditions(
+                [
+                    (
+                        self.algorithm in supported_algs,
+                        "Only lloyd algorithm is supported, elkan is computed using lloyd",
+                    ),
+                    (correct_count, "n_clusters is smaller than number of samples"),
+                    (
+                        np.allclose(sample_weight, np.ones_like(sample_weight)),
```

Comment:
This is weird. This is an unnecessary O(n) evaluation no matter what due to _check_sample_weight. If X is large, this could be a non-negligible impact on performance (on the order of an assert_all_finite call). There should only be sample weight checking elsewhere if it isn't None. Maybe put it with the _check_sample_weight behind an if sample_weight: and then have the evaluation of a constant value there (not just necessarily ones.

Reply:
good point, done. checking ones because we don't support sample_weight